src/v/iceberg/values.cc

```diff
@@ -0,0 +1,387 @@
+    template<typename T, typename U>
+    bool operator()(const T&, const U&) const {
+        static_assert(!std::is_same<T, U>::value);
+        return false;
+    }
+    template<typename T>
+    requires requires(T t) { t.val; }
+    bool operator()(const T& lhs, const T& rhs) const {
+        return lhs.val == rhs.val;
+    }
+};
+
+bool operator==(const primitive_value& lhs, const primitive_value& rhs) {
+    return std::visit(primitive_value_comparison_visitor{}, lhs, rhs);
+}
+
+bool operator==(const struct_value& lhs, const struct_value& rhs) {
+    if (lhs.fields.size() != rhs.fields.size()) {
+        return false;
+    }
+    for (size_t i = 0; i < lhs.fields.size(); i++) {
+        auto has_lhs = lhs.fields[i] != std::nullopt;
+        auto has_rhs = rhs.fields[i] != std::nullopt;
+        if (has_lhs != has_rhs) {
+            return false;
+        }
+        if (!has_lhs) {
+            // Both are null.
+            continue;
+        }
+        if (*lhs.fields[i] != *rhs.fields[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+bool operator==(
+  const std::unique_ptr<struct_value>& lhs,
+  const std::unique_ptr<struct_value>& rhs) {
+    if ((lhs == nullptr) != (rhs == nullptr)) {
+        return false;
+    }
+    if (lhs == nullptr) {
+        // Both null.
+        return true;
+    }
+    return *lhs == *rhs;
+}
+
+bool operator==(const list_value& lhs, const list_value& rhs) {
+    if (lhs.elements.size() != rhs.elements.size()) {
+        return false;
+    }
+    for (size_t i = 0; i < lhs.elements.size(); i++) {
+        auto has_lhs = lhs.elements[i] != std::nullopt;
+        auto has_rhs = rhs.elements[i] != std::nullopt;
+        if (has_lhs != has_rhs) {
+            return false;
+        }
+        if (!has_lhs) {
+            // Both are null.
+            continue;
+        }
+        if (*lhs.elements[i] != *rhs.elements[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+bool operator==(
+  const std::unique_ptr<list_value>& lhs,
+  const std::unique_ptr<list_value>& rhs) {
+    if ((lhs == nullptr) != (rhs == nullptr)) {
+        return false;
+    }
+    if (lhs == nullptr) {
+        // Both null.
+        return true;
+    }
+    return *lhs == *rhs;
+}
+
+bool operator==(const kv_value& lhs, const kv_value& rhs) {
+    auto has_lhs_val = lhs.val != std::nullopt;
+    auto has_rhs_val = rhs.val != std::nullopt;
+    if (has_lhs_val != has_rhs_val) {
+        return false;
+    }
+    if (lhs.key != rhs.key) {
+        return false;
+    }
+    if (has_lhs_val && *lhs.val != *rhs.val) {
+        return false;
+    }
+    return true;
+}
+
+bool operator==(const map_value& lhs, const map_value& rhs) {
+    if (lhs.kvs.size() != rhs.kvs.size()) {
+        return false;
+    }
+    for (size_t i = 0; i < lhs.kvs.size(); i++) {
+        if (lhs.kvs[i] != rhs.kvs[i]) {
+            return false;
+        }
+    }
+    return true;
+}
```

Comment:
i dunno how all this is used, but i guess the ordering of the map contents in the map container (vector-like) affects equality and hashing?

Reply:
Nice callout. Responded to a similar callout from Jim https://github.com/redpanda-data/redpanda/pull/22920#discussion_r1722012625